### PR TITLE
Support optional capacity on creation

### PR DESCRIPTION
**Pull request by fenollp**

### CODE REVIEW

**Comment by wk8** [comment]
Thanks for the PR @fenollp , might adding a test too?

**Comment by SVilgelm** [comment]
I think it would be great to have more general solution for the options.
something like 

```go

type Option struct {
  capacity int
}

func mergeOptions(opts ...Option) Option {
  var res Option
  for _, opt := range opts {
    if opt.capacity > 0 {
      res.capacity = opt.capacity
    }
  }
  return res
}

func WithCapacity(v int) Option {
  return Option{capacity: v}
}

func New[K comparable, V any](opts ...Option) *OrderedMap[K, V] {
  opt := mergeOptions(opts...)
  pairs = make(map[K]*Pair[K, V], opt.capacity)
  ...
}
```

this is more flexible solution, the option structure can be easily extended

**Comment by SVilgelm** [comment]
or using the functions based approach:
```go

type Option func(opt *option)

type option struct {
  capacity int
}

func mergeOptions(opts ...Option) option {
  res := &option{}
  for _, opt := range opts {
    opt(res)
  }
  return res
}

func WithCapacity(v int) Option {
  return func(opt *option) {
    opt.capacity = v
  }
}

func New[K comparable, V any](opts ...Option) *OrderedMap[K, V] {
  opt := mergeOptions(opts...)
  pairs = make(map[K]*Pair[K, V], opt.capacity)
  ...
}
```

**Comment by wk8** [comment]
@SVilgelm : I agree in general, but in this specific case:
1. it's nice to have an API that stays as close as possible to that of maps
2. I can't really think off the top of my head of any other options that could make sense in the future

So I'm inclined to keep this as is? Thoughts?

File: orderedmap_test.go [hunk 1]
```diff
@@ -263,6 +263,19 @@ func TestMove(t *testing.T) {
 	assert.NotEqual(t, err, nil)
 }
 
+func TestNewEdgeCases(t *testing.T) {
```
Comment (wk8) [inline orderedmap_test.go:266]: Nit: `TestNewWithCapacity` better

File: orderedmap_test.go [hunk 2]
```diff
@@ -263,6 +263,19 @@ func TestMove(t *testing.T) {
...
+	zero := New[int, string](0)
+	assert.Empty(t, zero.Len())
+
+	assert.PanicsWithValue(t, "too many arguments to New[K,V]()", func() {
+		_ = New[int, string](1, 2, 3)
+	})
+
+	om := New[int, string](-1)
```
Comment (wk8) [inline orderedmap_test.go:274]: I'm actually surprised that this doesn't panic...? That goes straight to `make(map[K]*Pair[K, V], capacity[0])`, and make does panic when given a < 0 capacity?

File: orderedmap_test.go [hunk 3]
```diff
@@ -263,6 +263,19 @@ func TestMove(t *testing.T) {
...
+	om.Set(1337, "quarante-deux")
+	assert.Equal(t, 1, om.Len())
+}
```
Comment (wk8) [inline orderedmap_test.go:277]: Would be nice to add a test with a > 0 capacity that also checks that `om.pairs` has the expected capacity.

**Review by wk8 — CHANGES_REQUESTED** [review]
A few nits, plus needs rebase. Thanks :)

File: orderedmap_test.go (hunk 2 above)
Comment (fenollp) [reply orderedmap_test.go:274]: No panic, just `min(capacity, 0)` 
See https://go.dev/play/p/sFwJi7omFA9

**Comment by fenollp** [comment]
Rebased!

> So I'm inclined to keep this as is? Thoughts?

Agreed!

File: orderedmap_test.go (hunk 2 above)
Comment (wk8) [reply orderedmap_test.go:272]: Interesting. https://go.dev/play/p/TiqBeNAUqe- does panic. I'll have to look more into that one :)

File: orderedmap_test.go (hunk 3 above)
Comment (wk8) [reply orderedmap_test.go:275]: Interesting 👍

**Pull request merged**